src/Features/Core/Portable/DocumentationComments/CodeFixes/AbstractAddDocCommentNodesCodeFixProvider.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.DiagnosticComments.CodeFixes
+{
+    internal abstract class AbstractAddDocCommentNodesCodeFixProvider
+        <TXmlElementSyntax, TXmlNameAttributeSyntax, TXmlTextSyntax, TMemberDeclarationSyntax> : CodeFixProvider
+        where TXmlElementSyntax : SyntaxNode
+        where TXmlNameAttributeSyntax : SyntaxNode
+        where TXmlTextSyntax : SyntaxNode
+        where TMemberDeclarationSyntax : SyntaxNode
+    {
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public async sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var parentMethod = root.FindNode(context.Span)?.FirstAncestorOrSelf<TMemberDeclarationSyntax>();
+
+            if (parentMethod != null)
+            {
+                context.RegisterCodeFix(
+                    new MyCodeAction(
+                        c => AddParamTagAsync(context.Document, context.Span, c)),
+                    context.Diagnostics);
+            }
+        }
+
+        protected abstract string NodeName { get; }
```

Comment:
Really, I only use `MemberDeclarationSyntax` to find the parent node that the doc comment will be on.  In the typeparam version, typeparam nodes are available on `TypeDeclarationSyntax` nodes, and `MethodDeclarationSyntax` nodes, if I am remembering anything, which could be tricky to find the correct parent--maybe this should be off-loaded to the language-specific implementation.  Then, the syntax factory needs to create a typeparam node instead of a param node, and those are the main differences.  Everything else should work.  However, as this is currently implemented, it relies on the feature of an already-existing node for the inspection--I could make it look for any doc comment node and try to add them after the `summary` node if there are no nodes of its specific type...  Or maybe the specific node type(s) to add them after should be offloaded to the non-abstract implementation because param should go after typeparam, then summary, then first, while typeparam nodes should either be first or after the summary node.  Either way, I think it would be much cleaner to keep this as two separate features.